### DISCUSSION TITLE
Adding WebSocket Specific Methods to IGraphQLClient

### DISCUSSION THREAD

**Discussion by ben-voss** [graphql-dotnet/graphql-client#526] [Ideas]
I'm looking at exposing the following GraphQLHttpClient specific members to the IGraphQLClient interface:

- WebSocketReceiveErrors
- WebsocketConnectionState
- InitialiseWebsocketConnection

This is so users can hide the concrete GraphQLHttpClient and use IGraphQLClient interface throughout their code bases making it easier to replace with mocks for unit testing.

Having tried to make an PR for this I found that the GraphQLWebsocketConnectionState is defined in the Abstractions.Websocket project making it inaccessible to the Abstractions project.  Moving types around will break backwards compatibility so the only way I can see to resolve this is to create a new IGraphQLWebSocketClient interface that extends IGraphQLClient and defines the extra methods.  GraphQLHttpClient would then implement both interfaces.

Before I try this, is this a good idea? Does anyone have a better suggestion?

**Comment by rose-a** [comment]
Hi, 

`IGraphQLClient` is meant to be transport-agnostic (since GraphQL is, too). This means you don't necessarily have a websocket behind it (imagine a `RS232GraphQLClient`).

You could propably put such an interface into *GraphQL.Client.Abstractions.Websocket* or straight into *GraphQL.Client* (which should have been named *GraphQL.Client.Http*, but that name on nuget.org has already been taken by someone...).

The only benefit I can see of having that interface and not deriving a class from `GraphQLHttpClient` is propably the use case you mentioned (mocking it in unit tests).

**Comment by ben-voss** [comment]
Thanks for the suggestions.  I have created a PR along these lines here:  https://github.com/graphql-dotnet/graphql-client/pull/537
